### PR TITLE
Implement omit_response_on_success batch option

### PR DESCRIPTION
**Pull request by teldosas**
This adresses #438 implementing the interface proposal of @SammyK . I haven't updated the docs yet. I will do so after the review.

### CODE REVIEW

File: src/Facebook/FacebookBatchRequest.php [hunk 1]
```diff
@@ -250,7 +257,10 @@ public function requestEntityToBatchArray(FacebookRequest $request, $requestName
             $batch['attached_files'] = $attachedFiles;
         }
 
-        // @TODO Add support for "omit_response_on_success"
+        if (isset($options['omit_response_on_success'])) {
```
Comment (SammyK) [inline src/Facebook/FacebookBatchRequest.php]: I'm thinking we just merge the options straight in so we can support all the existing options plus future ones.

```php
if (!empty($options)) {
    $batch = array_merge($options, $batch);
}
```

...or `array_merge($batch, $options)`, but that would allow the users to break things if they passed in an option named `headers` with random stuff in it for example.

But this makes me think we could get rid of `$requestName` and just pass in `name` through `$options`. Hum..

@yguedidi thoughts? :)

Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: I think `name`, `attached_files` and current `options` can be merge in one `options` parameter.
And to resolve your issue with merging @SammyK: `$batch += $options;` ;)

**Comment by teldosas** [comment]
@SammyK  @yguedidi I've made the changes :)

File: src/Facebook/Facebook.php [hunk 2]
```diff
@@ -483,6 +483,27 @@ public function sendBatchRequest(array $requests, $accessToken = null, $graphVer
     }
 
     /**
+     * Instantiates an empty FacebookBatchRequest entity.
+     *
+     * @param  AccessToken|string|null $accessToken  The top-level access token. Requests with no access token
+     *                                               will fallback to this.
+     * @param  string|null $graphVersion The Graph API version to use.
```
Comment (yguedidi) [inline src/Facebook/Facebook.php]: wrong description alignment :)
BTW, I'd add an empty line between this `@param` and the below `@return`

File: src/Facebook/FacebookBatchRequest.php [hunk 3]
```diff
@@ -62,16 +62,17 @@ public function __construct(FacebookApp $app = null, array $requests = [], $acce
     }
 
     /**
-     * A a new request to the array.
+     * Adds a new request to the array.
      *
      * @param FacebookRequest|array $request
      * @param string|null           $name
+     * @param array                 $options The batch options e.g. 'omit_response_on_success'
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: I'd typehint it as `string[]`

File: src/Facebook/FacebookBatchRequest.php [hunk 4]
```diff
@@ -91,6 +92,10 @@ public function add($request, $name = null)
             'request' => $request,
         ];
 
+        if (!empty($options)) {
+            $requestToAdd['options'] = $options;
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: I'd move this in the array initialization above, no need to check for empty, it will always be an array thanks to the typehint

File: src/Facebook/FacebookBatchRequest.php [hunk 5]
```diff
@@ -189,8 +194,19 @@ public function convertRequestsToJson()
     {
         $requests = [];
         foreach ($this->requests as $request) {
-            $attachedFiles = isset($request['attached_files']) ? $request['attached_files'] : null;
-            $requests[] = $this->requestEntityToBatchArray($request['request'], $request['name'], $attachedFiles);
+            $options = [];
+
+            if (isset($request['name'])) {
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: `name` is always set, I'd check for `!empty($request['name'])`

File: src/Facebook/FacebookBatchRequest.php [hunk 6]
```diff
@@ -216,12 +232,12 @@ public function validateBatchRequestCount()
      * Converts a Request entity into an array that is batch-friendly.
      *
      * @param FacebookRequest $request       The request entity to convert.
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: wrong description alignment

File: src/Facebook/FacebookBatchRequest.php [hunk 7]
```diff
@@ -189,8 +194,19 @@ public function convertRequestsToJson()
...
+
+            if (isset($request['name'])) {
+                $options['name'] = $request['name'];
+            }
+
+            if (isset($request['attached_files'])) {
+                $options['attached_files'] = $request['attached_files'];
+            }
+
+            $options += isset($request['options']) ? $request['options'] : [];
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: when my above comment about always set the `options` key, this become just `$options += $request['options'];` :)

File: src/Facebook/FacebookBatchRequest.php [hunk 8]
```diff
@@ -189,8 +194,19 @@ public function convertRequestsToJson()
...
         foreach ($this->requests as $request) {
-            $attachedFiles = isset($request['attached_files']) ? $request['attached_files'] : null;
-            $requests[] = $this->requestEntityToBatchArray($request['request'], $request['name'], $attachedFiles);
+            $options = [];
+
+            if (isset($request['name'])) {
+                $options['name'] = $request['name'];
+            }
+
+            if (isset($request['attached_files'])) {
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: like the `options` key, I'll always set the `attached_files` key in `add()`, and then check here for `!empty($request['attached_files'])`

**Comment by teldosas** [comment]
@yguedidi I made these changes too! :) The last review comment was not outdated even though I did force push and fixed the alignment. I'll start with the docs asap. ;)

**Comment by teldosas** [comment]
BTW `options`  and `attached_files` being set always cause 4 tests to fail because they use the method `getRequests()` which returns the `requests` array right after being processed from the `add()` method and the array contains two extra keys with empty or null values.

Also checking `!empty($request['name'])` before setting the name does not add the name of the firsts request when the requests are added as a numeric array because `0` is considered to be empty.

File: src/Facebook/FacebookBatchRequest.php [hunk 9]
```diff
@@ -189,8 +191,19 @@ public function convertRequestsToJson()
     {
         $requests = [];
         foreach ($this->requests as $request) {
-            $attachedFiles = isset($request['attached_files']) ? $request['attached_files'] : null;
-            $requests[] = $this->requestEntityToBatchArray($request['request'], $request['name'], $attachedFiles);
+            $options = [];
+
+            if (!empty($request['name'])) {
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: To solve the issues, check here for `null !== $request['name']`, and in the `add` function, cast the `$key` to a string :)

File: src/Facebook/FacebookBatchRequest.php [hunk 10]
```diff
@@ -189,8 +191,19 @@ public function convertRequestsToJson()
...
+                $options['name'] = $request['name'];
+            }
+
+            if (!empty($request['attached_files'])) {
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: BTW, you can change this too to `null !== $request['attached_files']` as `extractFileAttachments` always returns a string or null

**Comment by yguedidi** [comment]
Thanks @teldosas! I added some comments to fix your issues. About tests, just change the expected value in thoses failing tests :)

File: src/Facebook/FacebookBatchRequest.php (hunk 10 above)
Comment (teldosas) [inline src/Facebook/FacebookBatchRequest.php]: @yguedidi Then why not just use the `isset()` function that returns false when the variable is null. Plus I think that this is how we check in similar cases throughout the project.

File: src/Facebook/FacebookBatchRequest.php (hunk 9 above)
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: Updated, I'd added using `strval()`

File: src/Facebook/FacebookBatchRequest.php (hunk 10 above)
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: @teldosas "`extractFileAttachments` always returns a **non empty** string or null", sorry I missed an important part :)

`isset()` will of course work, but sementically it does not make sense, because the `attached_files` key is always set.
BTW, I tend to be strict about types: `null` means no attached files/no name, a string value means I want a value. 

Then the question is what to do when that value is invalid (e.g. an empty string).

I think we have two options :
- throwing an exception
- let Facebook Graph API returns an error

I personally prefer the last option. It's tedious to handle **all** user's type/value errors, we already pretty well document the usage of our functions :)

ping @SammyK, thoughts?

File: src/Facebook/FacebookBatchRequest.php (hunk 9 above)
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: @teldosas I'm talking about the `$key` variable in `add()`, the case when you give an array of requests.
In `->add($myRequest)`, `$name` will be null
In `->add($ManyRequests)`, `$name` for eachrequest will be its key in the array, an integer, which is not allowed by our doc, that's why I suggest you to `->add($request, strval($key))` in the loop :)

Comment (teldosas) [inline src/Facebook/FacebookBatchRequest.php]: Ok it makes sense now, thanks :)

**Comment by yguedidi** [comment]
Just one more thing, this PR targets master, which is the v6 dev branch, and this makes sense as it introduce a small BC break.
But it introduce a great addition that can be added without BC break, so what about closing this PR and open a new one that targets 5.4 and remove the BC break? :)

ping @SammyK @teldosas (sorry for making you wait again)

**Comment by teldosas** [comment]
I added the docs and I have some notes/questions:
1. After playing a bit around with this new feature I realized that this PR does not add support for `JSONP callback`. Furthermore I don't think that the **`JSONP callback`** can or should be implemented inside the `requestEntityToBatchArray` or the `add` method because it is **general option of every request**.
For example in a simple request one would add a `JSONP callback` like this: 
     ```php
     $req = $fb->get('/me/feed?callback'={function_name});
     ```
     (just by adding a `callback` get/post parameter).
But in the case of a `FacebookBatchRequest` we don't have anyway of adding a form post parameter. Thus I propose that we open a new issue to discuss an interface for adding a `JSONP callback` to a batch request. Maybe add a method 
     ```php 
     FacebookBatchRequest::setJSONPCallback(string $functionName)
     ```
     or even make it more general
     ```php 
     FacebookRequest::setJSONPCallback(string $functionName)
     ```
    Thoughts?

2. @yguedidi I am totally fine with the new PR but you would have to enlighten me about what this small Backwards Compatibility break is (I have no idea). :)

3. Is it a problem that we provide two (or maybe even three) ways for a user to create and send a batch request and only one of those supports the batch options (the one implemented in this PR)?

File: docs/examples/batch_request.md [hunk 11]
```diff
@@ -96,7 +96,130 @@ My next 2 events are House Warming Party,Some Foo Event.
...
+
+// Populate the batch request
+$batch->add($requestPostToFeed, "post-to-feed");
+
+// Set the 'depends_on' property to point to the first request
+$batch->add($requestLastPost, "last-post", ["depends_on" => "post-to-feed"]);
+
+// Send the batch request
+try {
+  $responses = $fb->getClient()->sendBatchRequest($batch);
```
Comment (teldosas) [inline docs/examples/batch_request.md]: Since this part that sends and prints the batch request is repeated in the four examples of this file, could we maybe add a new section here about sending a batch request? Maybe with two subsections:
* one for sending an array of `Facebook\FacebookRequest`s
* and one for sending an instance of `Facebook\FacebookBatchRequest`

Comment (yguedidi) [inline docs/examples/batch_request.md]: I agree, but this should be in a separate PR ;)

File: src/Facebook/FacebookBatchRequest.php [hunk 12]
```diff
@@ -216,12 +229,12 @@ public function validateBatchRequestCount()
...
      * @param FacebookRequest $request       The request entity to convert.
-     * @param string|null     $requestName   The name of the request.
-     * @param string|null     $attachedFiles Names of files associated with the request.
+     * @param array           $options       The name and the attached files of the request plus
+     *                                       the batch options e.g. 'omit_response_on_success'
      *
      * @return array
      */
-    public function requestEntityToBatchArray(FacebookRequest $request, $requestName = null, $attachedFiles = null)
+    public function requestEntityToBatchArray(FacebookRequest $request, array $options = [])
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: Here is the BC break, because `requestEntityToBatchArray` is public, changing its signature is a BC break. :)

**Comment by yguedidi** [comment]
@teldosas Thanks for the doc, excellent!

1. Why do you want to use JSONP in a PHP SDK? Your PHP code will not execute the JSONP response :)
2. See my comment in the code for the BC break. BTW, if you want to learn more about what breaks, Symfony framework has a good doc about it: http://symfony.com/doc/current/contributing/code/bc.html#working-on-symfony-code
3. Not for me, as we will work on v6. @SammyK thoughts?

**Comment by SammyK** [comment]
@teldosas Thanks for all the hard work on this! :)

1 & 2: +1 to what @yguedidi said :)
3: Let's implement this way first that supports all the options and if we want to create a facade on the `Facebook` class make batch requests easier, we can, but let's do that in another issue/PR. :)

File: src/Facebook/FacebookBatchRequest.php [hunk 13]
```diff
@@ -62,20 +62,21 @@ public function __construct(FacebookApp $app = null, array $requests = [], $acce
...
      * @param FacebookRequest|array $request
      * @param string|null           $name
+     * @param string[]              $options The batch options e.g. 'omit_response_on_success'
      *
      * @return FacebookBatchRequest
      *
      * @throws \InvalidArgumentException
      */
-    public function add($request, $name = null)
+    public function add($request, $name = null, array $options = [])
```
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: BTW, this is a BC break too in case someone extends `FacebookBatchRequest`

Comment (SammyK) [inline src/Facebook/FacebookBatchRequest.php]: Yeah, I'm curious if we can even add this to 5.4 without adding a BC break. Can you think of a way outside of adding a new method for "adding with options" or something?

Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: It looks a bit tricky, but it works! Renaming `$name` to `$options` and document that if it's a string it will be the request name, otherwise it will be the array of options :)

Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: @SammyK BC is something hard to keep, thats why we have to have the closest public API for v6, it ease adding features without break things :)

Comment (SammyK) [inline src/Facebook/FacebookBatchRequest.php]: Love it! Oh the wonky things we can get away with thanks to duck typing... :)

Comment (teldosas) [inline src/Facebook/FacebookBatchRequest.php]: So do we merge this to master and start a new PR for 5.4 with the aforementioned interface, or do we discard this PR and do just the PR for 5.4?

Also can I rename this to `Add full batch support` ?

File: src/Facebook/FacebookBatchRequest.php [hunk 14]
```diff
@@ -62,20 +62,21 @@ public function __construct(FacebookApp $app = null, array $requests = [], $acce
...
     {
         if (is_array($request)) {
             foreach ($request as $key => $req) {
-                $this->add($req, $key);
+                $this->add($req, strval($key));
```
Comment (SammyK) [inline src/Facebook/FacebookBatchRequest.php]: @teldosas Why did we add `strval()` here?
@yguedidi This is a BC break when you're expecting numeric array keys since they will be returned as strings now.

**Comment by SammyK** [comment]
> So do we merge this to master and start a new PR for 5.4 with the aforementioned interface, or do we discard this PR and do just the PR for 5.4?

I'd say let's close this PR and do a new PR that targets the 5.4 branch. Then I can merge it into master. Once merged into master, we can create a new PR with any BC breaks we want to do. :)

> Also can I rename this to Add full batch support ?

Sure!

> (About the JSONP) It makes sense. This @TODO tag was confusing ;)

Yeah, I remember adding that a while ago and then forgetting to remove it. :)

File: src/Facebook/FacebookBatchRequest.php (hunk 14 above)
Comment (yguedidi) [inline src/Facebook/FacebookBatchRequest.php]: @SammyK I tell @teldosas to add this. `add()` expects a string for the name, let's give it a string :)
How it can break?

**Comment by teldosas** [comment]
I started the PR for 5.4 :) #713

**Comment by SammyK** [comment]
Close since this was addressed in #713. Thanks! :)
